### ⚙️ **Mini-Practice**

1. Take a small dataset (10 numbers).
2. Compute Range, IQR, Variance, SD, and Mean Deviation.
3. Add an extreme value → watch Range & Variance increase sharply.
4. Compare IQR and Mean Deviation → see how they stay more stable.

In [2]:
import pandas as pd

data = {
    "Name": ["Aisha Khan", "Rohit Sharma", "Priya Verma", "Arjun Singh", "Sneha Patel",
             "Karan Mehta", "Meera Joshi", "Sameer Reddy", "Tanya Gupta", "Varun Das",
             "Anjali Sharma", "Rakesh Kumar", "Isha Kapoor", "Nikhil Jain", "Ritu Verma",
             "Deepak Singh", "Pooja Nair", "Aditya Roy", "Simran Kaur", "Mohit Bansal"],
    "Gender": ["F", "M", "F", "M", "F", "M", "F", "M", "F", "M", "F", "M", "F", "M", "F", "M", "F", "M", "F", "M"],
    "Age": [28, 32, 26, 30, 29, 35, 31, 27, 33, 29, 25, 34, 28, 31, 27, 30, 29, 33, 32, 26],
    "Salary": [48000, 55000, 45000, 60000, 50000, 65000, 52000, 47000, 58000, 51000,
               43000, 62000, 49000, 56000, 46000, 54000, 50000, 60000, 57000, 45000],
    "Dept": ["HR", "IT", "Marketing", "IT", "Finance", "IT", "Marketing", "HR", "Finance", "IT",
             "Marketing", "Finance", "HR", "IT", "Marketing", "Finance", "HR", "IT", "Finance", "Marketing"]
}

df = pd.DataFrame(data)

In [ ]:
# 1. Take a small dataset (10 numbers).

# 2. Compute Range, IQR, Variance, SD, and Mean Deviation.
print("2. Compute Range, IQR, Variance, SD, and Mean Deviation.")
Salary_Range =  max(df["Salary"]) -  min(df["Salary"])

print(df["Salary"].std())
print(Salary_Range)  # 22000

salary = df["Salary"].sort_values()

Q1 = salary.quantile(0.25)
Q3 = salary.quantile(0.75)
IQR = Q3 - Q1

print("Q1:", Q1)
print("Q3:", Q3)
print("IQR:", IQR)


# 3. Add an extreme value → watch Range & Variance increase sharply.
print("\n\nAdd an extreme value → watch Range & Variance increase sharply.\t Added:salary[0] = 12365479")
salary[0] = 12365479
Salary_Range1 =  max(salary) -  min(salary)

print(salary.std())  # standard deviation
print(Salary_Range1)  # Range

# 4. Compare IQR and Mean Deviation → see how they stay more stable.

Q1 = salary.quantile(0.25)
Q3 = salary.quantile(0.75)
IQR = Q3 - Q1

print("Q1:", Q1)
print("Q3:", Q3)
print("IQR:", IQR)




2. Compute Range, IQR, Variance, SD, and Mean Deviation.
6293.480335779409
22000
Q1: 47750.0
Q3: 57250.0
IQR: 9500.0


Add an extreme value → watch Range & Variance increase sharply.	 Added:salary[0] = 12365479
2753184.5146739567
12322479
Q1: 48500.0
Q3: 58500.0
IQR: 10000.0


Exactly — your output confirms that perfectly.

Here’s the breakdown from your results:

| Measure                | Before Outlier | After Outlier | Change     | Sensitivity         |
| ---------------------- | -------------- | ------------- | ---------- | ------------------- |
| **Range**              | 22,000         | 12,322,479    | Huge jump  | Extremely sensitive |
| **Standard Deviation** | –              | 2,753,184.51  | Huge       | Extremely sensitive |
| **IQR**                | 9,500          | 10,000        | Small rise | Stable              |

**Interpretation:**

* **Range and SD** explode when one extreme value is added.
  They depend on *every* number, so an outlier skews them badly.
* **IQR** changed only slightly, because quartiles depend on the **middle 50%** of data, not the extremes.

If you calculate **Mean Deviation**, you’ll see a similar result — small change compared to SD.

So yes, your test *proves statistically* that:

* Range and Standard Deviation are **outlier-sensitive**.
* IQR and Mean Deviation are **robust** and better for datasets with extreme values.


In [12]:
# Q1: 47750.0
# Q3: 57250.0
# IQR: 9500.0
print(df["Salary"].std())   # standard deviation
print(df["Salary"].var())   # variance
df["Salary"].quantile([0.25, 0.75])

6293.480335779409
39607894.7368421


0.25    47750.0
0.75    57250.0
Name: Salary, dtype: float64